pkg/compiler/global_test.go

```diff
@@ -80,6 +81,536 @@ func TestUnusedGlobal(t *testing.T) {
+					return a
+				}`
+		eval(t, src, big.NewInt(1), []interface{}{opcode.INITSSLOT, []byte{2}}, // sslot for A, B
+			opcode.PUSH2, opcode.STSFLD0, // store A
+			opcode.PUSH3, opcode.LDSFLD0, opcode.PUSH1, opcode.PUSH3, opcode.PACK, opcode.PUSH1, opcode.PICKITEM, opcode.STSFLD1, // evaluate B
+			opcode.PUSH1, []interface{}{opcode.CALL, []byte{8}}, opcode.LDSFLD1, opcode.ADD, opcode.DROP, opcode.RET, // evalute C and drop
+			opcode.PUSH1, opcode.RET, // Main
+			[]interface{}{opcode.INITSLOT, []byte{0, 1}}, opcode.LDARG0, opcode.RET) // f
+	})
+	t.Run("index expression", func(t *testing.T) {
+		src := `package foo
+				var Unused = 2
+				var A = f(1) + []int{1, 2, 3}[Unused] // index expression
+				func Main() int {
+					return 1
+				}
+				func f(a int) int {
+					return a
+				}`
+		eval(t, src, big.NewInt(1), []interface{}{opcode.INITSSLOT, []byte{1}}, // sslot for Unused
+			opcode.PUSH2, opcode.STSFLD0, // store Unused
+			opcode.PUSH1, []interface{}{opcode.CALL, []byte{14}}, // call f(1)
+			opcode.PUSH3, opcode.PUSH2, opcode.PUSH1, opcode.PUSH3, opcode.PACK, opcode.LDSFLD0, opcode.PICKITEM, // eval index expression
+			opcode.ADD, opcode.DROP, opcode.RET, // eval and drop A
+			opcode.PUSH1, opcode.RET, // Main
+			[]interface{}{opcode.INITSLOT, []byte{0, 1}}, opcode.LDARG0, opcode.RET) // f(a)
+	})
+	t.Run("used via nested function calls", func(t *testing.T) { // used in non directly used function
+		src := `package foo
+			var A = 1
+			func Main() int {
+				return f()
+			}
+			func f() int {
+				return g()
+			}
+			func g() int {
+				return A
+			}`
+		eval(t, src, big.NewInt(1), []interface{}{opcode.INITSSLOT, []byte{1}}, // sslot for A
+			opcode.PUSH1, opcode.STSFLD0, opcode.RET, // store A
+			[]interface{}{opcode.CALL, []byte{3}}, opcode.RET, // Main
+			[]interface{}{opcode.CALL, []byte{3}}, opcode.RET, // f
+			opcode.LDSFLD0, opcode.RET) // g
+	})
+	t.Run("struct field name matches global var name", func(t *testing.T) {
+		src := `package foo
+				type CustomStr struct { Int int	}
+				var str = CustomStr{Int: 2}
+				var Int = 5 // Unused and the code must not be emitted.
+				func Main() int {
+					return str.Int
+				}`
+		eval(t, src, big.NewInt(2), []interface{}{opcode.INITSSLOT, []byte{1}}, // sslot for str
+			opcode.PUSH2, opcode.PUSH1, opcode.PACKSTRUCT, opcode.STSFLD0, opcode.RET, // store str
+			opcode.LDSFLD0, opcode.PUSH0, opcode.PICKITEM, opcode.RET) // Main
+	})
+	t.Run("var as a struct field initializer", func(t *testing.T) {
+		src := `package foo
+				type CustomStr struct { Int int	}
+				var A = 5
+				var Int = 6 // Unused
+				func Main() int {
+					return CustomStr{Int: A}.Int
+				}`
+		eval(t, src, big.NewInt(5))
+	})
+	t.Run("argument of globally called function", func(t *testing.T) {
+		src := `package foo
+				var Unused = 5
+				var _, A = f(Unused)
+				func Main() int {
+					return 1
+				}
+				func f(int) (int, int) {
+					return 1, 2
+				}`
+		eval(t, src, big.NewInt(1))
```

Comment:
How can this test fail?

Reply:
The compilation may easily fail. I've found one bug thanks to this test, so I'd prefer to keep it. I've added `control` variable to make it more verbose.